### ISSUE TITLE
Handling of output parameter for worklet process()?

### ISSUE THREAD

**Issue by rtoy** [WebAudio/web-audio-api#1515]
Let's say you have a very simple `AudioWorklet` having a single input and a single output with no `AudioParams`.  Let's say make a mistake and have

```
process(inputs, outputs, params) {
  let input = inputs[0];
  let output = outputs[0];
  output.fill(input[0][0]);
  return true;
}
```
This accidentally modifies `outputs[0]` to be an array of floats.  But it's supposed to be a `sequence<Float32Array>`.

How is the worklet supposed to handle this?  I'm kind of thinking this should cause an error to be thrown.  I guess that also implies that if `output` isn't exactly `sequence<sequence<Float32Array>>` an error is thrown.  Similarly, if the length of the `Float32Array` isn't 128, it's an error.

(In case it matters, I actually did this by accident.)

**Comment by padenot** [comment]
> I'm kind of thinking this should cause an error to be thrown. I guess that also implies that if output isn't exactly sequence<sequence<Float32Array>> an error is thrown. Similarly, if the length of the Float32Array isn't 128, it's an error.

I agree.

**Comment by hoch** [comment]
Agreed - this is a spec oversight.

**Comment by hoch** [comment]
To clarify a bit:

1. Everything has to match for `sequence<sequence<Float32Array>>`.
2. Otherwise it throws an exception and deactivates the processor resulting the silence in the subsequent render quanta.
3. The corresponding AudioWorkletNode's `onprocessorerror` event handler will be fired.

**Comment by rtoy** [comment]
Yes, and the sequence lengths and Float32Array lengths must match.

**Comment by hoch** [comment]
See: https://github.com/w3c/css-houdini-drafts/issues/433

The problem is the error handling of WorkletGlobalScope is not well-defined yet.

**Comment by rtoy** [comment]
Actually, as pointed out to me by @sletz, I think we can solve this
immediate problem using `Object.freeze()` (See
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze)

Before calling `process()`, we can do something like:

```
inputs.forEach(in => Object.freeze(in))
Object.freeze(inputs)
```

This makes it so that the length of `inputs` can't be changed; the
elements of `inputs` can't be changed, and `inputs[k]` has fixed
length and the elements of `inputs[k]` are fixed to be `Float32Arrays`
of length 128.  You can, however, modify the elements of the
`Float32Arrays`.

And do the same thing with `outputs`.

If this is correct, then we don't have a problem with the user
trashing these parameters by accident---it just can't happen.

**Comment by hoch** [comment]
I like this idea a lot. The `Object.freeze` is [well-defined](https://tc39.github.io/ecma262/#sec-object.freeze), so we can use it in our spec or implementation.

**Comment by rtoy** [comment]
@chrisguttandin mentioned on the webaudio slack channel that `Object.freeze` isn't used commonly and the "general assumption in the JS community is that it is very slow and therefore should not be used in production code whenever possible"

I don't quite understand why it should be slow for our use case, unless it makes accessing the underlying Float32Arrays slow.

**Comment by hoch** [comment]
We can do the experiments and see how it actually performs. Just saw the guts of `Object.freeze` and the code is pretty lengthy.

**Comment by rtoy** [comment]
While we're waiting for the experiments, and assuming they work, we could say something like

1. Apply operation [Object.freeze ](https://tc39.github.io/ecma262/#sec-object.freeze) to `outputs`
1. For each n, apply freeze to `outputs[n]`

Simple experiments shows that this works. `outputs` is always `sequence<sequence<Float32Array>>`. (Well, for my test `Array<Array<Float32Array>>`.)

**Comment by karlt** [comment]
At some point, it would be nice to let the AudioWorkletProcessor determine the
number of output channels.  This would be useful for nodes that have a tail
time but wish to adjust the output channel count according to the
computedNumberOfChannels from the input.  Consider a DelayNode implementation,
for example.

Could we permit process() to modify the length of outputs?

I guess freeze could still be applied to outputs[n], but it would kind of lose its appeal somewhat.

**Comment by rtoy** [comment]
Teleconf:  Just say changing it is undefined and move @karlt comment https://github.com/WebAudio/web-audio-api/issues/1515#issuecomment-406173462 to a new issue, for v.next.

**Comment by rtoy** [comment]
To clarify a bit from the teleconf, for V1, we will leave the behavior undefined if you modify the output parameter.  In v.next, we can take up the issue about allowing the user to change the length of the array.  Leaving it undefined now allows us to define the behavior in a compatible way for v.next.

**Comment by karlt** [comment]
`AudioNode.channelCount` specifies the number of input channels, and so would not be appropriate for specifying the number of output channels.

**Comment by hoch** [comment]
I have some doubts on what `channelCount` means for AudioNode. For the sake of simplicity, the spec says it is about the input, but you can see it in many different ways. From my perspective, the channel count means a number of channels being processed by a "kernel" inside of an AudioNode container.

For example, when you set a `channelCount` of 5 for a GainNode, the node configures the kernel so it can process 5 channels. The input and the output of the node will be configured accordingly afterwards. So you'll end up with a 5-channel output. Then does this `channelCount` really mean "input channels"?

Also this is why I don't like the idea of having a separate `outputChannelCount` property on AudioWorkletNode. It feels like adding more confusion to a thing that's already ambiguous.

**Comment by hoch** [comment]
I think the mutable output array will be a significant architectural change for any browser.

I believe we (as in all browsers) configure AudioNode's input and output first and then process the audio. For Audio Worklet, this will change such model so we can 1) configure input, 2) process audio and then 3) configure output. I'll have to go back to take a look when the downstream propagation happens, but as @rtoy suggested the change will not be simple and might have other implication we did not expect.

**Comment by padenot** [comment]
I think this is easy for us, @karlt will know for sure.

**Comment by karlt** [comment]
Yes, Gecko already has a model where output is configured after audio input is received for processing, and so no major changes would be required.  IOW Gecko can easily set the output channel count after process() returns.

**Comment by hoch** [comment]
Unfortunately Blink does the channel count configuration at the beginning/end of the render quantum. So I think we'll have a different behavior on this one. To accommodate both cases, we can say something like "the channel count will change immediately or at the next render quantum."

**Comment by hoch** [comment]
This is technically a sub issue of #1780. WG decided to punt this to v.next.

**Comment by rtoy** [comment]
An additional note from today's teleconf:  We'd like to gather up all enhancements suggested for an AudioWorklet and consider them all in a unified way in the next version.  We want to have a relatively stable definition of AudioWorklet now since we're nearing recommendation status.

**Comment by karlt** [comment]
A, perhaps iterable, interface type with an indexed property getter is an option for at least the inner objects, of length number of channels, to control which modifications are permitted.  This could permit modifications to the length, but exclude modifications to the type of the properties, if that is desired, for example.

FrozenArray would be an option for the outer object `outputs`, of length number of outputs, assuming we don't want the inner objects replaced with other objects.

There seem to be varying opinions on whether frozen objects are desirable or not, but they are in WebIDL in support of APIs that provide the same object each time. See for example https://www.w3.org/Bugs/Public/show_bug.cgi?id=23682#c34 or discussion in https://www.w3.org/Bugs/Public/show_bug.cgi?id=29004
